### PR TITLE
Py38 update

### PR DESCRIPTION
**Pull request by mshriver**
Should be merged after @JacobCallahan PR #7924 which introduces py3.8.

This enforces it, as well as uses pyupgrade tool to apply python 3.8 language standards automatically via pre-commit.

Initial f-string conversions were done by pyupgrade. It doesn't convert all uses of `.format()`

Real question: is anyone ever going to actually review a 326 file PR?

### CODE REVIEW

File: robottelo/api/utils.py [hunk 1]
```diff
@@ -440,8 +430,8 @@ def configure_provisioning(org=None, loc=None, compute=False, os=None):
             .search(
                 query={
                     'search': 'family="Redhat" '
-                    'AND major="{0}" '
-                    'AND minor="{1}")'.format(
+                    'AND major="{}" '
+                    'AND minor="{}")'.format(
                         os.split(' ')[1].split('.')[0], os.split(' ')[1].split('.')[1]
                     )
                 }
```
Comment (JacobCallahan) [inline robottelo/api/utils.py:437]: This looks pretty ugly still, and could still look better as an f-string. Perhaps something like this
```python
os_ver = os.split(' ')[1].split('.')
query={'search': 'family="Redhat" AND major="{os_ver[0]}" AND minor="{os_ver[1]}")'}
```

File: robottelo/api/utils.py [hunk 2]
```diff
@@ -647,39 +637,35 @@ def wait_for_syncplan_tasks(repo_backend_id=None, timeout=10, repo_name=None):
...
         # It is '[]' string for empty content when backend_identifier is wrong
         if len(req.content) > 2:
             if req.json()[0].get('state') in ['finished']:
                 return True
             elif req.json()[0].get('error'):
                 raise AssertionError(
-                    "Pulp task with repo_id {0} errored or not "
-                    "found: '{1}'".format(repo_backend_id, req.json().get('error'))
+                    "Pulp task with repo_id {} errored or not "
+                    "found: '{}'".format(repo_backend_id, req.json().get('error'))
```
Comment (JacobCallahan) [inline robottelo/api/utils.py:668]:
```suggestion
                    f"Pulp task with repo_id {repo_backend_id} errored or not "
                    f"found: '{req.json().get('error')}'"
```

File: robottelo/cli/base.py [hunk 3]
```diff
@@ -144,8 +144,8 @@ def _handle_response(cls, response, ignore_stderr=None):
         """
         if response.return_code != 0:
             full_msg = (
-                'Command "{0} {1}" finished with return_code {2}\n'
-                'stderr contains following message:\n{3}'.format(
+                'Command "{} {}" finished with return_code {}\n'
+                'stderr contains following message:\n{}'.format(
                     cls.command_base, cls.command_sub, response.return_code, response.stderr
                 )
```
Comment (JacobCallahan) [inline robottelo/cli/base.py:150]:
```suggestion
                f'Command "{cls.command_base} {cls.command_sub}" finished with return_code '
                f'{response.return_code}\nstderr contains following message:\n{response.stderr}' 
```

**Review by JacobCallahan — CHANGES_REQUESTED** [review]

File: robottelo/libvirt_discovery.py [hunk 4]
```diff
@@ -94,7 +94,7 @@ def __init__(
         self.ip_addr = None
         self._domain = None
         self._created = False
-        self.guest_name = 'mac{0}'.format(self.mac.replace(':', ""))
+        self.guest_name = 'mac{}'.format(self.mac.replace(':', ""))
```
Comment (JacobCallahan) [inline robottelo/libvirt_discovery.py:97]:
```suggestion
        self.guest_name = f"""mac{self.mac.replace(':', "")}"""
```

File: tests/foreman/cli/test_activationkey.py [hunk 5]
```diff
@@ -1511,12 +1510,12 @@ def test_positive_view_subscriptions_by_non_admin_user(self):
...
-        ak_name_like = 'ak_{0}'.format(gen_string('alpha'))
+        ak_name_like = 'ak_{}'.format(gen_string('alpha'))
         hc_names_like = (
-            'Test_*_{0}'.format(gen_string('alpha')),
-            'Test_*_{0}'.format(gen_string('alpha')),
+            'Test_*_{}'.format(gen_string('alpha')),
+            'Test_*_{}'.format(gen_string('alpha')),
         )
-        ak_name = '{0}_{1}'.format(ak_name_like, gen_string('alpha'))
+        ak_name = '{}_{}'.format(ak_name_like, gen_string('alpha'))
```
Comment (JacobCallahan) [inline tests/foreman/cli/test_activationkey.py:1518]: you know the thing

File: tests/foreman/cli/test_capsule.py [hunk 6]
```diff
@@ -56,11 +55,11 @@ def test_negative_create_with_url(self):
 
         :CaseLevel: Component
 
-       """
+        """
         # Create a random proxy
         with self.assertRaisesRegex(CLIFactoryError, 'Could not create the proxy:'):
             make_proxy(
-                {'url': 'http://{0}:{1}'.format(gen_string('alpha', 6), gen_string('numeric', 4))}
+                {'url': 'http://{}:{}'.format(gen_string('alpha', 6), gen_string('numeric', 4))}
```
Comment (JacobCallahan) [inline tests/foreman/cli/test_capsule.py:62]:
```suggestion
                {'url': f'http://{gen_string("alpha", 6)}:{gen_string("numeric", 4)}'}
```

File: tests/foreman/cli/test_capsule_installer.py [hunk 7]
```diff
@@ -186,7 +185,7 @@ def test_positive_reinstall_on_same_node_after_remove(self):
                 Capsule.refresh_features({'name': capsule_vm._capsule_hostname})
             # katello-remove is no longer part of product so deploy it sideways to perform testing
             capsule_vm.run(
-                "wget -nv -P /usr/bin http://{0}/pub/katello-remove".format(
+                "wget -nv -P /usr/bin http://{}/pub/katello-remove".format(
                     settings.artifacts_server
                 )
```
Comment (JacobCallahan) [inline tests/foreman/cli/test_capsule_installer.py:190]:
```suggestion
                f"wget -nv -P /usr/bin http://{settings.artifacts_server}/pub/katello-remove"
```

File: tests/foreman/cli/test_capsule_installer.py [hunk 8]
```diff
@@ -203,11 +202,11 @@ def test_positive_reinstall_on_same_node_after_remove(self):
...
             self.assertEqual(result.return_code, 0)
             # generate capsule certs and installer command
-            cert_file_path = '/tmp/{0}-certs.tar'.format(capsule_vm.hostname)
+            cert_file_path = f'/tmp/{capsule_vm.hostname}-certs.tar'
             result = ssh.command(
                 'capsule-certs-generate '
-                '--foreman-proxy-fqdn {0} '
-                '--certs-tar {1}'.format(capsule_vm.hostname, cert_file_path)
+                '--foreman-proxy-fqdn {} '
+                '--certs-tar {}'.format(capsule_vm.hostname, cert_file_path)
```
Comment (JacobCallahan) [inline tests/foreman/cli/test_capsule_installer.py:209]: i think this reads better
```suggestion
                f'--foreman-proxy-fqdn {capsule_vm.hostname} '
                f'--certs-tar {cert_file_path}'
```

File: tests/foreman/cli/test_classparameters.py [hunk 9]
```diff
@@ -50,12 +49,12 @@ def setUpClass(cls):
...
         be able to work with unique entity for each specific test.
         """
-        super(SmartClassParametersTestCase, cls).setUpClass()
+        super().setUpClass()
         cls.puppet_modules = [{'author': 'robottelo', 'name': 'cli_test_classparameters'}]
         cls.org = make_org()
         cls.loc = make_location()
         cv = publish_puppet_module(cls.puppet_modules, CUSTOM_PUPPET_REPO, cls.org['id'])
-        cls.env = Environment.list({'search': 'content_view="{0}"'.format(cv['name'])})[0]
+        cls.env = Environment.list({'search': 'content_view="{}"'.format(cv['name'])})[0]
```
Comment (JacobCallahan) [inline tests/foreman/cli/test_classparameters.py:57]:
```suggestion
        cls.env = Environment.list({'search': f"""content_view="{cv['name']}""""})[0]
```

File: tests/foreman/cli/test_classparameters.py [hunk 10]
```diff
@@ -69,7 +68,7 @@ def setUpClass(cls):
         cls.sc_params_list = SmartClassParameter.list(
             {
                 'environment': cls.env['name'],
-                'search': 'puppetclass="{0}"'.format(cls.puppet_class['name']),
+                'search': 'puppetclass="{}"'.format(cls.puppet_class['name']),
```
Comment (JacobCallahan) [inline tests/foreman/cli/test_classparameters.py:71]: could follow the same format as above

File: tests/foreman/cli/test_classparameters.py [hunk 11]
```diff
@@ -194,7 +193,7 @@ def test_positive_import_twice_list_by_puppetclass_id(self):
         :CaseImportance: Low
         """
         cv = publish_puppet_module(self.puppet_modules, CUSTOM_PUPPET_REPO, self.org['id'])
-        env = Environment.list({'search': 'content_view="{0}"'.format(cv['name'])})[0]
+        env = Environment.list({'search': 'content_view="{}"'.format(cv['name'])})[0]
```
Comment (JacobCallahan) [inline tests/foreman/cli/test_classparameters.py:196]: and this

File: tests/foreman/cli/test_computeresource_libvirt.py [hunk 12]
```diff
@@ -59,8 +58,8 @@ def valid_name_desc_data():
         {'name': gen_string('alphanumeric'), 'description': gen_string('alphanumeric', 255)},
         {'name': gen_string('utf8'), 'description': gen_string('utf8')},
         {
-            'name': '<html>{0}</html>'.format(gen_string('alpha')),
-            'description': '<html>{0}</html>'.format(gen_string('alpha')),
+            'name': '<html>{}</html>'.format(gen_string('alpha')),
+            'description': '<html>{}</html>'.format(gen_string('alpha')),
```
Comment (JacobCallahan) [inline tests/foreman/cli/test_computeresource_libvirt.py:62]: can bring these in

File: tests/foreman/cli/test_computeresource_libvirt.py [hunk 13]
```diff
@@ -127,7 +126,7 @@ def test_positive_create_with_name(self):
         """
         ComputeResource.create(
             {
-                'name': 'cr {0}'.format(gen_string(str_type='alpha')),
+                'name': 'cr {}'.format(gen_string(str_type='alpha')),
```
Comment (JacobCallahan) [inline tests/foreman/cli/test_computeresource_libvirt.py:129]:
```suggestion
                'name': f'cr {gen_string(str_type="alpha")}',
```

File: tests/foreman/cli/test_computeresource_rhev.py [hunk 14]
```diff
@@ -60,7 +59,7 @@ def test_positive_create_rhev_with_valid_name(self):
         """
         ComputeResource.create(
             {
-                'name': 'cr {0}'.format(gen_string(str_type='alpha')),
+                'name': 'cr {}'.format(gen_string(str_type='alpha')),
```
Comment (JacobCallahan) [inline tests/foreman/cli/test_computeresource_rhev.py:62]: more of the same throughout this file

File: tests/foreman/cli/test_contentview.py [hunk 15]
```diff
@@ -2935,7 +2935,7 @@ def test_positive_sub_host_with_restricted_user_perm_at_custom_loc(self):
             host_client.install_katello_ca()
             host_client.register_contenthost(
                 org['label'],
-                lce='{0}/{1}'.format(env['name'], content_view['name']),
+                lce='{}/{}'.format(env['name'], content_view['name']),
```
Comment (JacobCallahan) [inline tests/foreman/cli/test_contentview.py:2938]:
```suggestion
                lce=f'{env["name"]}/{content_view["name"]}',
```

File: tests/foreman/cli/test_contentview.py [hunk 16]
```diff
@@ -3087,7 +3087,7 @@ def test_positive_sub_host_with_restricted_user_perm_at_default_loc(self):
...
```
Comment (JacobCallahan) [inline tests/foreman/cli/test_contentview.py:3090]: ditto

File: tests/foreman/cli/test_docker.py [hunk 17]
```diff
@@ -1249,7 +1249,7 @@ def test_positive_pull_image(self):
             # publishing takes few seconds sometimes
             result, _ = wait_for(
                 lambda: ssh.command(
-                    'docker pull {0}'.format(repo['published-at']),
+                    'docker pull {}'.format(repo['published-at']),
```
Comment (JacobCallahan) [inline tests/foreman/cli/test_docker.py:1252]: these are throughout this file
```suggestion
                    f'docker pull {repo["published-at"]}',
```

File: tests/foreman/cli/test_domain.py [hunk 18]
```diff
@@ -37,7 +36,7 @@ def valid_create_params():
     """Returns a list of valid domain create parameters"""
     return [
         {
-            'name': 'white spaces {0}'.format(gen_string(str_type='utf8')),
+            'name': 'white spaces {}'.format(gen_string(str_type='utf8')),
```
Comment (JacobCallahan) [inline tests/foreman/cli/test_domain.py:39]: and this file

File: tests/foreman/cli/test_environment.py [hunk 19]
```diff
@@ -43,15 +42,15 @@ class EnvironmentTestCase(CLITestCase):
...
+        super().setUpClass()
         cls.org = entities.Organization().create()
         cls.loc = entities.Location().create()
         cls.loc2 = entities.Location().create()
 
         # Setup for puppet class related tests
         puppet_modules = [{'author': 'robottelo', 'name': 'generic_1'}]
         cls.cv = publish_puppet_module(puppet_modules, CUSTOM_PUPPET_REPO, cls.org.id)
-        cls.env = Environment.list({'search': 'content_view="{0}"'.format(cls.cv['name'])})[0]
+        cls.env = Environment.list({'search': 'content_view="{}"'.format(cls.cv['name'])})[0]
```
Comment (JacobCallahan) [inline tests/foreman/cli/test_environment.py:53]: you guessed it!

**Review by JacobCallahan — CHANGES_REQUESTED** [review]
Just a few things. :)

**Comment by mshriver** [comment]
@JacobCallahan there by my search (after having done more conversions manually) over 600 remaining uses of `.format()` in robottelo.

I don't really intend to completely remove the use of `.format()`, it still has its purpose and appropriate places to use it.

Thoughts on my limiting the amount of time I'm spending directly updating all of these, and leaving their conversion to future contributions? People will be refactoring their tests, and defining new fixtures, and along the way we can continue to remove uses of .format where f-strings make more sense.

Going to push what I've updated just searching directly for .format, not necessarily addressing the specific comments you've made here to places where pyupgrade did not automatically change to f-string.

**Comment by JacobCallahan** [comment]
@mshriver it sounds like it is time to update and further expand SatQE's code standards. I can take some examples from this pull request for that as well. We can also raise this updated standard to the rest of the reviewers and then the larger team.

I'm fine with the limited scope of your changes in this pull request. Will look more at it tomorrow morning.

**Comment by mshriver** [comment]
@JacobCallahan great - I did include an update to the docs where string formatting is mentioned, stating to use f-string instead of .format.

File: robottelo/cli/base.py [hunk 20]
```diff
@@ -288,12 +286,12 @@ def execute(
...
-        cmd = 'LANG={0} {1} hammer -v {2} {3} {4} {5}'.format(
+        cmd = 'LANG={} {} hammer -v {} {} {} {}'.format(
             settings.locale,
             'time -p' if time_hammer else '',
-            '-u {0}'.format(user) if user is not None else '--interactive no',
-            '-p {0}'.format(password) if password is not None else '',
-            '--output={0}'.format(output_format) if output_format else '',
+            f'-u {user}' if user is not None else '--interactive no',
+            f'-p {password}' if password is not None else '',
+            f'--output={output_format}' if output_format else '',
```
[the remark this answers is not in the document]
Comment (mshriver) [reply robottelo/cli/base.py]: I need to roll this back, this broke the unit tests. Functionality isn't the same, as argument keys are getting included when they should not.

**Review by JacobCallahan — APPROVED** [review]

**Review by pondrejk — APPROVED** [review]
It's too huge to make any guarantees, though I don't know how else to do it and now is probably the best time

**Pull request merged**